src/commands/build-util.ts

```diff
@@ -14,74 +15,81 @@ export function getInputsFromOutput(fileLoc: string, plugins: SnowpackPlugin[])
+ * functon and return it's result.
+ * 
+ * If no match is found, fall back to just readign 
+ * the file from disk and return it.
+ */
+async function runPipelineLoadStep(
   srcPath: string,
-  srcContents: string,
   {buildPipeline, messageBus, isDev}: BuildFileOptions,
-): Promise<SnowpackBuildMap> {
+) {
   const srcExt = getExt(srcPath).baseExt;
-  const rootFileName = path.basename(srcPath).replace(srcExt, '');
-  const output: SnowpackBuildMap = {
-    [srcExt]: srcContents,
-  };
-
-  // Run the file through the Snowpack build pipeline:
-  for (const step of buildPipeline as SnowpackPluginWithLegacy[]) {
-    // DEPRECATED: Plugin legacy transform() support. If a plugin defines a transform() function,
-    // call it. Transform cannot change the file extension, and was designed to run on every file
-    // type and return null if no change was needed.
-    if (step.transform) {
-      for (const destExt of Object.keys(output)) {
-        const destBuildFile = output[destExt];
-        const result = await step.transform({
-          contents: destBuildFile,
-          fileExt: destExt,
-          filePath: rootFileName + destExt,
-          isDev,
-          log: (msg, data = {}) => {
-            messageBus.emit(msg, {
-              ...data,
-              id: step.name,
-              msg: data.msg && `[${srcPath}] ${data.msg}`,
-            });
-          },
-          // Deprecated
-          urlPath: `./${path.basename(rootFileName + destExt)}`,
+  for (const step of buildPipeline) {
+    if (!step.resolve || !step.resolve.input.includes(srcExt)) {
+      continue;
+    }
+    if (!step.load) {
+      continue;
+    }
+    const result = await step.load({
+      fileExt: srcExt,
+      filePath: srcPath,
+      isDev,
+      log: (msg, data = {}) => {
+        messageBus.emit(msg, {
+          ...data,
+          id: step.name,
+          msg: data.msg && `[${srcPath}] ${data.msg}`,
         });
-        if (result) {
-          output[destExt] = result.result;
-        }
-      }
+      },
+    });
+    const mainOutputExt = step.resolve.output[0];
+    if (typeof result === 'string') {
+      return {[mainOutputExt]: result};
+    } else if (result && typeof result === 'object' && result.result) {
+      return {[mainOutputExt]: result.result};
+    } else if (result && typeof result === 'object') {
+      return result;
+    } else {
       continue;
     }
+  }
+  return {[srcExt]: await fs.readFile(srcPath, getEncodingType(srcExt))};
+}
 
-    // For now, we only run through build plugins that match at least one file extension.
-    // All other plugins are ignored.
-    if (!step.build || step.input.length === 0) {
+/**
+ * Build Plugin Second Pass: If a plugin defines a
+ * transform() method,call it. Transform cannot change
+ * the file extension, and was designed to run on
+ * every file type and return null/undefined if no
+ * change needed.
+ */
+async function runPipelineTransformStep(
```

Comment:
👍 Good clarification and comments here. I feel this is clear to understand from reading the source